### PR TITLE
[IPU] update ipu_backend

### PR DESCRIPTION
**Pull request by gglin001**
### PR types

Others
### PR changes

Others
### Describe

1. support setting weight_decay for single tensors
2. avoid memcpy for feed_op when using ipu
3. building paddle with [popdist](https://docs.graphcore.ai/projects/poprun-user-guide/en/latest/introduction.html)
4. add `ipu_backend.weights_to_host()` for syncing states from ipu

### CODE REVIEW

File: paddle/fluid/framework/tensor_util.cc [hunk 1]
```diff
@@ -447,27 +458,15 @@ void TensorCopySync(const Tensor& src, const platform::Place& dst_place,
...
-    PADDLE_THROW(platform::errors::Unimplemented(
-        "Copy from %s to %s is not supported.", src_place, dst_place));
-  }
-#endif
 #ifdef PADDLE_WITH_CUSTOM_DEVICE
   else if (platform::is_custom_place(src_place) &&  // NOLINT
            platform::is_cpu_place(dst_place)) {     /* custom_device -> cpu*/
     memory::Copy(dst_place, dst_ptr, src_place, src_ptr, size, nullptr);
-  }
+  }                                                // NOLINT
```
Comment (gglin001) [inline paddle/fluid/framework/tensor_util.cc:465]: 这里和下面的几个 `// NOLINT` 是为了维持现有的格式, 否则 clang-format 检查会 fail

File: paddle/fluid/operators/controlflow/feed_op.cc [hunk 2]
```diff
@@ -40,6 +40,10 @@ class FeedVariableVisitor : public boost::static_visitor<void> {
         out_var_->GetMutable<framework::LoDTensor>();
     if (platform::is_same_place(in_tensor.place(), place_)) {
       out_tensor->ShareDataWith(in_tensor);
+#ifdef PADDLE_WITH_IPU
+    } else if (platform::is_ipu_place(place_)) {
+      out_tensor->ShareDataWith(in_tensor);
+#endif
```
Comment (qili93) [inline paddle/fluid/operators/controlflow/feed_op.cc:49]: 这里加一点注释说明一下IPU的tensor place的处理情况？比如IPU下的in_tensor place还是CPUPlace, 因为PopART会自动处理host和device至今的数据拷贝，因此IPU的tensor在Paddle中全部是CPU tensor

**Review by qili93 — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/controlflow/feed_op.cc (hunk 2 above)
Comment (gglin001) [reply paddle/fluid/operators/controlflow/feed_op.cc:49]: fixed

**Review by qili93 — APPROVED** [review]
LGTM

**Pull request merged**